dnsprovider/aws.go

```diff
@@ -0,0 +1,315 @@
+	ChangeResourceRecordSets(*route53.ChangeResourceRecordSetsInput) (*route53.ChangeResourceRecordSetsOutput, error)
+	ListHostedZonesPages(input *route53.ListHostedZonesInput, fn func(resp *route53.ListHostedZonesOutput, lastPage bool) (shouldContinue bool)) error
+	ListHostedZonesByName(input *route53.ListHostedZonesByNameInput) (*route53.ListHostedZonesByNameOutput, error)
+	CreateHostedZone(*route53.CreateHostedZoneInput) (*route53.CreateHostedZoneOutput, error)
+	DeleteHostedZone(*route53.DeleteHostedZoneInput) (*route53.DeleteHostedZoneOutput, error)
+}
+
+// AWSProvider is an implementation of DNSProvider for AWS Route53.
+type AWSProvider struct {
+	Client Route53API
+	DryRun bool
+}
+
+// Zones returns the list of hosted zones.
+func (p *AWSProvider) Zones() ([]string, error) {
+	zones := []string{}
+
+	f := func(resp *route53.ListHostedZonesOutput, lastPage bool) (shouldContinue bool) {
+		for _, zone := range resp.HostedZones {
+			zones = append(zones, *zone.Name)
+		}
+
+		return true
+	}
+
+	err := p.Client.ListHostedZonesPages(&route53.ListHostedZonesInput{}, f)
+	if err != nil {
+		return zones, err
+	}
+
+	return zones, nil
+}
+
+// Zone returns a single zone given a DNS name.
+func (p *AWSProvider) Zone(dnsName string) (*route53.HostedZone, error) {
+	params := &route53.ListHostedZonesByNameInput{
+		DNSName: aws.String(dnsName),
+	}
+
+	resp, err := p.Client.ListHostedZonesByName(params)
+	if err != nil {
+		return nil, err
+	}
+
+	if len(resp.HostedZones) != 1 {
+		return nil, fmt.Errorf("not exactly one hosted zone found by name, got %d", len(resp.HostedZones))
+	}
+
+	return resp.HostedZones[0], nil
+}
+
+// CreateZone creates a hosted zone given a name.
+func (p *AWSProvider) CreateZone(name string) (*route53.HostedZone, error) {
+	params := &route53.CreateHostedZoneInput{
+		CallerReference: aws.String(uuid.New().String()),
+		Name:            aws.String(name),
+	}
+
+	resp, err := p.Client.CreateHostedZone(params)
+	if err != nil {
+		return nil, err
+	}
+
+	return resp.HostedZone, nil
+}
+
+// DeleteZone deletes a hosted zone given a name.
+func (p *AWSProvider) DeleteZone(name string) error {
+	params := &route53.DeleteHostedZoneInput{
+		Id: aws.String(name),
+	}
+
+	_, err := p.Client.DeleteHostedZone(params)
+	if err != nil {
+		return err
+	}
+
+	return nil
+}
+
+// Records returns the list of records in a given hosted zone.
+func (p *AWSProvider) Records(zone string) ([]endpoint.Endpoint, error) {
+	hostedZone, err := p.Zone(zone)
+	if err != nil {
+		return nil, err
+	}
+
+	params := &route53.ListResourceRecordSetsInput{
+		HostedZoneId: hostedZone.Id,
+	}
+
+	endpoints := []endpoint.Endpoint{}
+
+	f := func(resp *route53.ListResourceRecordSetsOutput, lastPage bool) (shouldContinue bool) {
+		for _, r := range resp.ResourceRecordSets {
+			if *r.Type != route53.RRTypeA {
```

Comment:
likewise everywhere. Should we use aws `IntVar, StringVar etc` functions for pointer dereferencing